scala-package/core/src/main/scala/org/apache/mxnet/api/java/ArgBuilder.scala

```diff
@@ -0,0 +1,69 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.mxnet.api.java
+
+import scala.collection.mutable
+import scala.collection.mutable.ListBuffer
+import collection.JavaConverters._
+
+/**
+  * This arg Builder is intent to solve Java to Scala conversion
+  * to take the input such as (arg: Any*)
+  */
+class ArgBuilder {
```

Comment:
why do we need this? why are you going back in time? we want to move away from (arg: Any*) to type-safe APIs

Reply:
I think type-safe is not as important as ease of usage. For type-safe in Java, the major question is how to deal with default args.

Reply:
I guess Java does not get defaults. @andrewfayres rightly asked how many APIs do we have that have a large number of parameters to matter? may be we just add builder to those if they are only a few and for others they can pass Scala Option.None or change the APIs to accept gauva Optional. 
@lanking520 can you find how many have more than 5 parameters?

Reply:
After running this:
```Scala
    printf(s"\n\n\n\nTotal numbers " +
      ndarrayFunctions.map(_.listOfArgs.length)
        .count(_ > 5)
      + " out of " + ndarrayFunctions.length + "\n\n\n\n"
    )
```
get output
```
Total numbers 64 out of 665
```

Reply:
It's not the total number of parameters that matters, it's how many default parameters there are in the method. If a method has 5 parameters and no defaults then the builder doesn't help any. If there are 5 parameters and all are defaults then it helps a lot.

@lanking520 Can we get a count of how many methods have more than 3 default args? If possible what I'd really like is a distribution (x methods have 1 default arg, y have 2, ...) but if this is too difficult I understand.

Reply:
```
printf(s"\n\n\n\nTotal numbers " +
      ndarrayFunctions.count(_.listOfArgs.count(_.isOptional) > 3)
      + " out of " + ndarrayFunctions.length + "\n\n\n\n"
    )
```
Here you go
```
Total numbers 65 out of 665
```

Reply:
So about 10% of NDArray methods have more than 3 default args.

I'm going to give some more thought before I commit to this but my initial reaction is that we include this but try not to promote it's use too much in docs/examples. Leave it there as an ease of use option for the customer with the understanding that when they use this they will be giving up type-safety.